sklearn/linear_model/sag_fast.pyx

```diff
@@ -32,62 +27,270 @@ cdef inline double fmax(double x, double y) nogil:
+
+        where:
+            prediction = dot(x_sample, weights) + intercept
+            \delta_{y,c} = 1 if (y == c) else 0
+            sw = sample_weight
+
+        Parameters
+        ----------
+        prediction : pointer to a np.ndarray[double] of shape (n_classes,)
+            Prediction of the multinomial classifier, for current sample.
+
+        y : double, between 0 and n_classes - 1
+            Indice of the correct class for current sample.
+
+        n_classes : integer
+            Total number of classes.
+
+        sample_weight : double
+            Weight of current sample.
+
+        Returns
+        -------
+        loss : double
+            Multinomial loss for current sample.
+
+        Reference
+        ---------
+        Bishop, C. M. (2006). Pattern recognition and machine learning.
+        Springer. (Chapter 4.3.4)
+        """
+        cdef double logsumexp_prediction = _logsumexp(prediction, n_classes)
+        cdef double loss
+
+        # y is the indice of the correct class of current sample.
+        loss = (logsumexp_prediction - prediction[int(y)]) * sample_weight
+        return loss
+
+    cdef void _dloss(self, double* prediction, double y, int n_classes,
+                     double sample_weight, double* gradient_ptr) nogil:
+        """Multinomial Logistic regression gradient of the loss.
+
+        The gradient of the multinomial logistic loss for one sample is:
+        grad_c = - sw * (p[c] - \delta_{y,c})
+
+        where:
+            p[c] = exp(logsumexp(prediction) - prediction[c])
+            prediction = dot(sample, weights) + intercept
+            \delta_{y,c} = 1 if (y == c) else 0
+            sw = sample_weight
+
+        Note that to obtain the true gradient, this value has to be multiplied
+        by the sample vector x.
+
+        Parameters
+        ----------
+        prediction : pointer to a np.ndarray[double] of shape (n_classes,)
+            Prediction of the multinomial classifier, for current sample.
+
+        y : double, between 0 and n_classes - 1
+            Indice of the correct class for current sample.
+
+        n_classes : integer
+            Total number of classes.
+
+        sample_weight : double
+            Weight of current sample.
+
+        gradient_ptr : pointer to a np.ndarray[double] of shape (n_classes,)
+            Gradient vector to be filled.
+
+        Reference
+        ---------
+        Bishop, C. M. (2006). Pattern recognition and machine learning.
+        Springer. (Chapter 4.3.4)
+        """
+        cdef double logsumexp_prediction = _logsumexp(prediction, n_classes)
+        cdef int class_ind
+
+        for class_ind in range(n_classes):
+            gradient_ptr[class_ind] = exp(prediction[class_ind] -
+                                          logsumexp_prediction)
+
+        # y is the indice of the correct class of current sample.
+        gradient_ptr[int(y)] -= 1.0
+
+        for class_ind in range(n_classes):
```

Comment:
You can do `gradient_ptr[int(y)] = -1` initially and do this loop along with the first, right?